RecoBTag/DeepFlavour/plugins/DeepFlavourTagInfoProducer.cc

```diff
@@ -0,0 +1,318 @@
+
+  for (std::size_t jet_n = 0; jet_n <  jets->size(); jet_n++) {
+
+    // create data containing structure
+    deep::DeepFlavourFeatures features;
+
+    // reco jet reference (use as much as possible)
+    const auto & jet = jets->at(jet_n);
+    // dynamical castoting to pointers, null if not possible
+    const auto * pf_jet = dynamic_cast<const reco::PFJet *>(&jet);
+    const auto * pat_jet = dynamic_cast<const pat::Jet *>(&jet);
+    edm::RefToBase<reco::Jet> jet_ref(jets, jet_n);
+    // TagInfoCollection not in an associative container so search for matchs
+    const edm::View<reco::ShallowTagInfo> & taginfos = *shallow_tag_infos;
+    edm::Ptr<reco::ShallowTagInfo> match;
+    // Try first by 'same index'
+    if ((jet_n < taginfos.size()) && (taginfos[jet_n].jet() == jet_ref)) {
+        match = taginfos.ptrAt(jet_n);
+    } else {
+      // otherwise fail back to a simple search
+      for (auto itTI = taginfos.begin(), edTI = taginfos.end(); itTI != edTI; ++itTI) {
+        if (itTI->jet() == jet_ref) { match = taginfos.ptrAt( itTI - taginfos.begin() ); break; }
+      }
+    }
+    reco::ShallowTagInfo tag_info;
+    if (match.isNonnull()) {
+      tag_info = *match; 
+    } // will be default values otherwise
+
+    // fill basic jet features
+    deep::jet_features_converter(jet, features.jet_features);
+
+    // fill number of pv
+    features.npv = vtxs->size();
+
+    // fill features from ShallowTagInfo
+    const auto & tag_info_vars = tag_info.taggingVariables();
+    btag_features_converter(tag_info_vars, features.tag_info_features);
+
+    // copy which will be sorted
+    auto svs_sorted = *svs;     
+    // sort by dxy
+    std::sort(svs_sorted.begin(), svs_sorted.end(),
+              [&pv](const auto & sva, const auto &svb)
+              { return deep::sv_vertex_comparator(sva, svb, pv); });
+    // fill features from secondary vertices
+    for (const auto & sv : svs_sorted) {
+      if (reco::deltaR(sv, jet) > jet_radius_) continue;
+      else {
+        features.sv_features.emplace_back();
+        // in C++17 could just get from emplace_back output
+        auto & sv_features = features.sv_features.back();
+        deep::sv_features_converter(sv, pv, jet, sv_features);
+      }
+    }
+
+    // stuff required for dealing with pf candidates 
+    math::XYZVector jet_dir = jet.momentum().Unit();
+    GlobalVector jet_ref_track_dir(jet.px(),
+                                   jet.py(),
+                                   jet.pz());
+
+    std::vector<sorting::sortingClass<size_t> > c_sorted, n_sorted;
+
+    deep::TrackInfoBuilder trackinfo(track_builder);
+
+    // unsorted reference to sv
+    const auto & svs_unsorted = *svs;     
+    // fill collection, from DeepTNtuples plus some styling
+    for (unsigned int i = 0; i <  jet.numberOfDaughters(); i++){
+        auto cand = dynamic_cast<const reco::Candidate *>(jet.daughter(i));
```

Comment:
is there an example where this dynamic_cast is actually necessary?
reco::Jet::daughter return type is a ```const reco::Candidate *```

Reply:
not needed, will remove